httpclient5/src/main/java/org/apache/hc/client5/http/async/methods/DeflatingGzipEntityProducer.java

```diff
@@ -0,0 +1,250 @@
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * ====================================================================
+ *
+ * This software consists of voluntary contributions made by many
+ * individuals on behalf of the Apache Software Foundation.  For more
+ * information on the Apache Software Foundation, please see
+ * <http://www.apache.org/>.
+ *
+ */
+package org.apache.hc.client5.http.async.methods;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.zip.CRC32;
+import java.util.zip.Deflater;
+
+import org.apache.hc.core5.http.Header;
+import org.apache.hc.core5.http.nio.AsyncEntityProducer;
+import org.apache.hc.core5.http.nio.DataStreamChannel;
+import org.apache.hc.core5.util.Args;
+
+/**
+ * Streams an {@link AsyncEntityProducer} through raw DEFLATE
+ * and wraps the result in a valid GZIP container.
+ * <p>
+ * Memory usage is bounded (8 KiB buffers) and back-pressure
+ * from the I/O reactor is honoured.
+ *
+ * @since 5.6
+ */
+public final class DeflatingGzipEntityProducer implements AsyncEntityProducer {
+
+    /* ---------------- constants & buffers --------------------------- */
+
+    private static final int IN_BUF = 8 * 1024;
+    private static final int OUT_BUF = 8 * 1024;
+
+    private final AsyncEntityProducer delegate;
+    private final CRC32 crc = new CRC32();
+    private final Deflater deflater = new Deflater(Deflater.DEFAULT_COMPRESSION, true);
+    private final byte[] in = new byte[IN_BUF];
+    private final ByteBuffer outBuf = ByteBuffer.allocate(OUT_BUF);
+
+    private boolean headerSent = false;
+    private boolean finished = false;
+    private long uncompressed = 0;
+
+    private final AtomicBoolean released = new AtomicBoolean(false);
+
+    public DeflatingGzipEntityProducer(final AsyncEntityProducer delegate) {
+        this.delegate = Args.notNull(delegate, "delegate");
+        outBuf.flip(); // start in “read mode” with no data
+    }
+
+    /* ------------------- metadata ------------------- */
+
+    @Override
+    public boolean isRepeatable() {
+        return false;
```

Comment:
@arturobernalg GZIP coding should be repeatable, if the underlying entity itself is repeatable, should it not?
`delegate.isRepeatable`